Auto spot price

ninadsp opened the pull request:
First pass at trying to get current spot price and then bidding 2 cents more than that.

Things that still need to happen:
* Throw exceptions that CDK can understand
* How do we prompt the user about the spot price that we calculated?
* Update sample config file and readme

Commit:
Implement auto spot pricing

SathyaBhat commented on requirements.txt:
```diff
@@ -1 +1,2 @@
+boto3
```
the CDK pattern seems to be have this in setup.py https://github.com/SathyaBhat/folding-aws/blob/master/setup.py#L21

not sure when/why this was done but lets keep it in one place

SathyaBhat commented on folding_aws/folding_asg_stack.py:
```diff
@@ -30,3 +39,27 @@ def __init__(self, scope: core.Construct, id: str, region: str, vpc: str, ec2_in
+      if len(all_azs) == 0:
+        print("Could not find any Availability Zone in region {}. \
+        Please check AWS configuration and whether this region is enabled for you".format(region))
+        sys.exit(1)
```
I don't see an import for `sys`

ninadsp replied:
I initially wrote this up in `app.py` and tested with it, but changed my mind at the last moment because that seemed like the better place for this. Copy/pasta error :(

SathyaBhat commented on folding_aws/folding_asg_stack.py:
```diff
@@ -30,3 +39,27 @@ def __init__(self, scope: core.Construct, id: str, region: str, vpc: str, ec2_in
+      all_azs_info = client.describe_availability_zones(Filters=[{'Name': 'region-name', 'Values': [region]}])
+      all_azs = [az['ZoneName'] for az in all_azs_info['AvailabilityZones']]
+      if len(all_azs) == 0:
+        print("Could not find any Availability Zone in region {}. \
```
since we're using Python 3.6+, I find f strings easier to read/use ``

```
print(f"Could not find any Availability Zone in {region} region.  Please check AWS configuration and whether this region is enabled for you")
```

note the `{region}` which references the variable

ninadsp commented:
I'm trying to reason about where is the best place to execute the 'compute spot price and ask the user about it'.

1. I originally started off by including it in `app.py`'s `main()`, but that was fine to test it. It doesn't feel like the right place because we're leaking abstractions (the CloudFormation stack doesn't need to care about the details of instance pricing).
2. I then moved it to `folding_asg_stack.py` and computed it before we define the ASG, which seems the right place for who should own the data. And, it is fine to prompt a user 'is this price acceptable for you? y/N' question here. Exiting if we get a no is manageable too. But, what happens on the second or third run? Do we compute the price every time and thus, end up making changes to the stack with every run? I don't see anything in the CDK documentation for 'hey, this is the first run, perform actions only now, don't ever do them again'.

Is there any other option available to handle this? Or, do we go back to option 1, and keep it as a sub command to just compute the spot price for the region+instance type we're interested in, and ask the user to manually set the values in the config file?

SathyaBhat commented:
> 
> 
> I'm trying to reason about where is the best place to execute the 'compute spot price and ask the user about it'.

How about doing the spot price checks in `configure()`? 

- First time run/no spot price configured: query boto API, present the price to the user and have them accept it/reject it. If rejected, we raise an error and stop right there.. If accepted, update config file
  - Subsequent runs/price in configure: still check for current spot price and if it is not the same, ask the user and have them accept/reject. If rejected, we raise an error and stop right there.. If accepted, update config file.
- Have a `--force` option will accept values in config/spot price in config and proceed without asking the user?

That way, everything is contained in `configure()`

SathyaBhat commented on app.py:
```diff
@@ -7,37 +7,41 @@
-  asg_stack = FoldingAsgStack(app, "folding-asg", region=config['aws']['region'], vpc=vpc_stack.vpc, 
+    app = core.App()
+    vpc_stack = FoldingVpcStack(app, "folding-vpc", env=core.Environment(region=config['aws']['region']))
+    asg_stack = FoldingAsgStack(app, "folding-asg", region=config['aws']['region'], vpc=vpc_stack.vpc,
```
since we're using different editors, I see the indentation has change, Might I suggest adding an `.editorconfig`?

https://editorconfig.org/

ninadsp replied:
Works for me! Do you want to check in your config file, so that I can rebase my branch and work with that?

ninadsp commented:
>     * First time run/no spot price configured:

There is no way to know whether this is the first run or not, unless we decide to put a file ourselves to track that this has been configured/run atleast once before hand. I'm not a big fan of that style though. This will just have to run every time.

SathyaBhat commented:
> 
> 
> > ```
> > * First time run/no spot price configured:
> > ```
> 
> There is no way to know whether this is the first run or not, unless we decide to put a file ourselves to track that this has been configured/run atleast once before hand. I'm not a big fan of that style though. This will just have to run every time.

yeah, thats fine by me

SathyaBhat commented:
@ninadsp is this ready to be merged?

ninadsp commented:
Rebased and squashed all commits. This branch should now query the AWS API in the selected region and fetch spot prices for the chosen instance type, during the CDK synth process. If no spot price is configured in the config file, or it is lower than the current spot prices, a spot price will be proposed and the user can accept it. If the user accepts, the price is also written to the config file.
